File: lib/easyfrontend/lyap.c
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, see <http://www.gnu.org/licenses/>.
//
// Copyright (C) Peter Benner, Martin Koehler, Jens Saak and others
//               2009-2018
//

/**
 * @file lib/easyfrontend/lyap.c
 * @brief Easy to use Lyapunov solver interface.
 * @author @koehlerm
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <math.h>
#include "mess/mess.h"
#include "mess/error_macro.h"
#include <complex.h>

/**
 * @brief Frontend to compute a factorized solution of a (generalized) Lyapunov Equation.
 * @param[in] A   input \f$ A \f$ matrix of the Lyapunov Equation
 * @param[in] E   input \f$ E \f$ matrix of the Lyapunov Equation \n
 *                (NULL if \f$ E \f$ does not exist)
 * @param[in] B    input right hand side of the Lyapunov Equation
 * @param[out] Z Cholesky factor \f$ Z \f$ of the solution \f$ X \approx ZZ^T \f$
 * @return zero on success or a non-zero error value otherwise
 *
 * The @ref mess_lyap function solves either
 * \f[
 *  AX +XA^T +BB^T =0
 * \f]
 * or
 * \f[
 *  AXE^T + EXA^T + BB^T = 0
 * \f]
 * depending on the existence of the matrix \f$ E \f$. \n
 * If the matrix object for \f$ E  \f$ is set to @c NULL the standard Lyapunov Equation is solved, otherwise the
 * generalized one is solved.\n
 * Depending on the storage type of the matrix \f$ A \f$ the solver decides whether a dense solver
 * \ref mess_direct_create_generalized_lyapunovchol or a sparse solver \ref mess_lrcfadi_adi is used. In both cases the solution is returned
 * as a Cholesky factor \f$ X \approx ZZ^T \f$.
 *
 * \attention
 * The \ref mess_lrcfadi_adi uses its default parameters set by \ref mess_options_init. \n
 * If these settings are not applicable to the given problem, the equation have to be solved by configuring the
 * ADI properly and call \ref mess_lrcfadi_adi separately.
 *
 */
int mess_lyap ( mess_matrix A, mess_matrix E, mess_matrix B, mess_matrix Z )
{
    int haveE = 0 ;
    MSG_FNAME(__func__);
    mess_matrix  Etmp= NULL ;
    int ret = 0 ;


    /*-----------------------------------------------------------------------------
     *  check input
     *-----------------------------------------------------------------------------*/
    mess_check_nullpointer(A);
    mess_check_nullpointer(B);
    mess_check_real(A);
    mess_check_real(B);
    mess_check_nullpointer(Z);

    if ( E != NULL ) {
        haveE=1;
        mess_check_same_size(A,E);
    }

    if ( MESS_IS_DENSE(A)) {
        /*-----------------------------------------------------------------------------
         *  solve a dense lyapunov eqn
         *-----------------------------------------------------------------------------*/
        if (haveE) {
            /*-----------------------------------------------------------------------------
             *  with mass matrix
             *-----------------------------------------------------------------------------*/
            if ( MESS_IS_DENSE(E)) {
                Etmp = E;
            } else {
                ret = mess_matrix_init(&Etmp);                                      FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_matrix_init);
                ret = mess_matrix_convert(E,Etmp,MESS_DENSE);                       FUNCTION_FAILURE_HANDLE(ret,(ret!=0), mess_matrix_convert);
            }

            mess_direct lyapsol;
            ret = mess_direct_init(&lyapsol);                                       FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_direct_init);
            ret = mess_direct_create_generalized_lyapunovchol(A, Etmp, lyapsol);    FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_direct_create_generalize_lyapunovchol);
            ret = mess_direct_solvem(MESS_OP_NONE, lyapsol, B, Z);                  FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_direct_solvem);
            mess_direct_clear(&lyapsol);

            if ( Etmp != E ) {
                mess_matrix_clear(&Etmp);
            }
        } else {
            /*-----------------------------------------------------------------------------
             *  standard case
             *-----------------------------------------------------------------------------*/
            mess_direct lyapsol;
            ret = mess_direct_init(&lyapsol);                                       FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_direct_init);
            ret = mess_direct_create_generalized_lyapunovchol(A,NULL,lyapsol);      FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_direct_create_lyapunovchol);
            ret = mess_direct_solvem(MESS_OP_NONE, lyapsol, B, Z);                  FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_direct_solvem);
            mess_direct_clear(&lyapsol);
        }

    } else {

        /*-----------------------------------------------------------------------------
         *  Solve with LRCFADI
         *-----------------------------------------------------------------------------*/
            int ok =0 ;
            mess_options opt;
            mess_status  stat;
            mess_equation     eqn;

            ret = mess_options_init(&opt);  FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_options_init);
            ret = mess_status_init(&stat);  FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_status_init);
            ret = mess_equation_init(&eqn);     FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_equation_init);
            opt->adi_res2_tol = MESS_MIN(1e-13 * sqrt(A->rows), sqrt(mess_eps()));
            opt->adi_res2c_tol =MESS_MIN(1e-14 * sqrt(A->rows), sqrt(mess_eps()));
            opt->adi_rel_change_tol = MESS_MIN(mess_eps() * A->rows, sqrt(mess_eps()));

            ret = mess_equation_lyap(eqn, opt, A, E, B);    FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_equation_lyap);

            ret = mess_parameter(eqn, opt, stat);           FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_parameter);
            ret = mess_lrcfadi_adi(eqn, opt, stat, Z);          FUNCTION_FAILURE_HANDLE(ret, (ret!=0), mess_lrcfadi_adi);

            ok = stat->stop_res2 || stat->stop_res2c || stat->stop_rel || stat->stop_user;
            if ( !ok ) {
                MSG_ERROR("The ADI did not converge. Please check your matrix or change the solver parameters.\n");
                mess_status_print(stat);
                mess_equation_clear(&eqn);
                mess_status_clear(&stat);
                mess_options_clear(&opt);
                return MESS_ERROR_CONVERGE;
            }
            mess_equation_clear(&eqn);
            mess_status_clear(&stat);
            mess_options_clear(&opt);
    }

    return 0;
}       /* -----  end of function mess_easy_lyap  ----- */

File: python/unittests/easy/lyap.py
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.
#
# Copyright (C) Peter Benner, Martin Koehler, Jens Saak and others
#               2009-2018
#

"""Test for lyap function."""
from __future__ import print_function, division
import unittest
from scipy.io import mmread
from scipy.sparse  import csc_matrix, csr_matrix, coo_matrix
import numpy as np
from pymess import res2_lyap, MESS_OP_NONE
import pymess

class Testlyap(unittest.TestCase):
    """Test for lyap function."""
    amtx = '@CMAKE_SOURCE_DIR@/tests/data/Rail/A.mtx'
    bmtx = '@CMAKE_SOURCE_DIR@/tests/data/Rail/B.mtx'
    emtx = '@CMAKE_SOURCE_DIR@/tests/data/Rail/E.mtx'
    matformats = [csc_matrix, csr_matrix, coo_matrix]

    def setUp(self):
        """is called before every test function"""
        self.a = mmread(self.amtx).tocsr()
        self.b = mmread(self.bmtx)
        self.e = mmread(self.emtx).tocsr()
        self.tol = 1e-10

    def template_lyap(self, hase):
        """template to reduce code in the test cases"""
        #solve lyapunov equation
        a = self.a
        e = self.e if hase else None
        b = self.b
        z = pymess.lyap(a, e, b)
        nrmr, nrmrhs, relnrm = res2_lyap(a, e, b, z, MESS_OP_NONE)
        print("res2 = {0:e}\t rel = {1:e}\t res2/rel = {2:e}".format(nrmr, nrmrhs, relnrm))
        self.assertLessEqual(relnrm, self.tol)

    def test_lyap_filter(self):
        """test"""
        for mf1 in self.matformats:
            for mf2 in self.matformats:
                self.a = mf1(self.a)
                self.e = mf2(self.e)
                self.template_lyap(False)

    def test_lyap_generalized_filter(self):
        """test"""
        for mf1 in self.matformats:
            for mf2 in self.matformats:
                self.a = mf1(self.a)
                self.e = mf2(self.e)
                self.template_lyap(True)

    def template_lyap_dense(self, n, hase):
        """template to reduce code in test cases"""

        # generate stable matrix A
        d = np.diag(np.ravel(-10 * np.random.rand(n, 1) - 10))
        q, _ = np.linalg.qr(np.random.rand(n, n))
        a = q.dot(d).dot(q.T)

        # generate B
        b = np.random.rand(n, np.max([1, np.int(n / 10)]))

        if hase:
            # generate matrix E
            e = np.diag(np.ravel(np.random.rand(n, 1)))

            # solve lyapunov
            z = pymess.lyap(a, e, b)

            # residual
            nrmr, nrmrhs, relnrm = res2_lyap(a, e, b, z, MESS_OP_NONE)

        else:
            # solve lyapunov
            z = pymess.lyap(a, None, b)

            # residual
            nrmr, nrmrhs, relnrm = res2_lyap(a, None, b, z, MESS_OP_NONE)

        print("res2 = {0:e}\t rel = {1:e}\t res2/rel = {2:e}".format(nrmr, nrmrhs, relnrm))
        self.assertLessEqual(relnrm, self.tol)

    def test_lyap_dense_1(self):
        """test"""
        print("\n")
        self.template_lyap_dense(1, 0)

    def test_lyap_dense_2(self):
        """test"""
        print("\n")
        self.template_lyap_dense(2, 0)

    def test_lyap_dense_5(self):
        """test"""
        print("\n")
        self.template_lyap_dense(5, 0)

    def test_lyap_dense_10(self):
        """test"""
        print("\n")
        self.template_lyap_dense(10, 0)

    def test_lyap_dense_100(self):
        """test"""
        print("\n")
        self.template_lyap_dense(100, 0)

    def test_lyap_dense_250(self):
        """test"""
        print("\n")
        self.template_lyap_dense(250, 0)

    def test_lyap_dense_333(self):
        """test"""
        print("\n")
        self.template_lyap_dense(333, 0)

    def test_glyap_dense_1(self):
        """test"""
        print("\n")
        self.template_lyap_dense(1, 1)

    def test_glyap_dense_2(self):
        """test"""
        print("\n")
        self.template_lyap_dense(2, 1)

    def test_glyap_dense_5(self):
        """test"""
        print("\n")
        self.template_lyap_dense(5, 1)

    def test_glyap_dense_10(self):
        """test"""
        print("\n")
        self.template_lyap_dense(10, 1)

    def test_glyap_dense_100(self):
        """test"""
        print("\n")
        self.template_lyap_dense(100, 1)

    def test_glyap_dense_250(self):
        """test"""
        print("\n")
        self.template_lyap_dense(250, 1)

    def test_glyap_dense_333(self):
        """test"""
        print("\n")
        self.template_lyap_dense(333, 1)

    def tearDown(self):
        """called after every test function"""
        pass

    def run(self, result=None):
        """ Stop after first error """
        if not result.errors:
            super(Testlyap, self).run(result)
